src/Util/ApiLoader.php

```diff
@@ -0,0 +1,229 @@
+            throw new RuntimeException($failureMessage . 'Response returned unsuccessfully');
+        }
+
+        // Allow callback to handle processing of the response body
+        $result = $callback($responseBody);
+
+        // Setting the value to the cache for subsequent requests
+        $this->handleCacheFromResponse($response, $result);
+
+        return $result;
+    }
+
+    /**
+     * @return Client
+     */
+    public function getGuzzleClient()
+    {
+        return $this->guzzleClient;
+    }
+
+    /**
+     * @param Client $guzzleClient
+     * @return $this
+     */
+    public function setGuzzleClient(Client $guzzleClient)
+    {
+        $this->guzzleClient = $guzzleClient;
+        return $this;
+    }
+
+    /**
+     * Attempts to load something from the cache and deserializes from JSON if successful
+     *
+     * @return array|bool
+     * @throws \Psr\SimpleCache\InvalidArgumentException
+     */
+    protected function getFromCache()
+    {
+        $cacheKey = $this->getCacheKey();
+        $result = $this->getCache()->get($cacheKey, false);
+        if ($result === false) {
+            return false;
+        }
+
+        // Deserialize JSON object and return as an array
+        return Convert::json2array($result);
+    }
+
+    /**
+     * Given a value, set it to the cache with the given key after serializing the value as JSON
+     *
+     * @param string $cacheKey
+     * @param mixed $value
+     * @param int $ttl
+     * @return bool
+     * @throws \Psr\SimpleCache\InvalidArgumentException
+     */
+    protected function setToCache($cacheKey, $value, $ttl = null)
+    {
+        // Seralize as JSON to ensure array etc can be stored
+        $value = Convert::raw2json($value);
+
+        return $this->getCache()->set($cacheKey, $value, $ttl);
+    }
+
+    /**
+     * Check the API response for cache control headers and respect them internally in the SilverStripe
+     * cache if found
+     *
+     * @param Response $response
+     * @param array|string $result
+     * @throws \Psr\SimpleCache\InvalidArgumentException
+     */
+    protected function handleCacheFromResponse(Response $response, $result)
+    {
+        // Handle caching if requested
+        if ($cacheControl = $response->getHeader('Cache-Control')) {
+            // Combine separate header rows
+            $cacheControl = implode(', ', $cacheControl);
+
+            if (strpos($cacheControl, 'no-store') === false
+                && preg_match('/(?:max-age=)(\d+)/i', $cacheControl, $matches)
+            ) {
+                $duration = (int) $matches[1];
+
+                $cacheKey = $this->getCacheKey();
+                $this->setToCache($cacheKey, $result, $duration);
+            }
+        }
+    }
+
+    /**
+     * @return CacheInterface
+     */
+    public function getCache()
+    {
+        if (!$this->cache) {
+            $this->cache = Injector::inst()->get(CacheInterface::class . '.supportedAddons');
```

Comment:
We should use a different namespace now this is an abstract class.

Reply:
Do you have a suggestion for a good replacement? Perhaps `BringYourOwnIdeas\Maintenance` for this class and + `\ApiLoader` for its subclasses?

Reply:
I meant cache namespace. Ha sorry for the confusion.

Reply:
Ah yep, true

Reply:
Probably just a generic cache namespace for all loaders would have been fine as each "loader" only uses one cache key but I don't think it really makes a difference.

Reply:
Yep! Updated